keps/sig-node/4216-image-pull-per-runtime-class/README.md

```diff
@@ -0,0 +1,1338 @@
+This would help differentiate different versions of the same image on a given node.
+
+### Containerd startup/LoadImages()
+
+LoadImages() is an API called during containerd startup to load all the exisiting images from containerd's store to CRI. client.ListImages() is called to get the list of images from containerd. At this point, we have no information of what runtime was used to pull the image. Therefore, before loading the image into CRI's store we need to find the approprite platfomr matcher to use by walking through criservice.platformMatcherMap() as follows:
+
+**Current:**
+
+<pre>
+
+// recover recovers system state from containerd and status checkpoint.
+func (c *criService) recover(ctx context.Context) error {
+...
+...
+	// Recover all images.
+	cImages, err := c.client.ListImages(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to list images: %w", err)
+	}
+	c.loadImages(ctx, cImages)
+...
+...
+}
+</pre>
+
+
+**Proposed**
+
+<pre>
+
+// recover recovers system state from containerd and status checkpoint.
+func (c *criService) recover(ctx context.Context) error {
+...
+...
+	// Recover all images.
+    <mark>
+	listImagesOpt := []containerd.ListImageOpt {
+		containerd.ListImageWithPlatformMatcherMap(c.platformMatcherMap),
+	}
+	cImages, err := c.client.ListImagesWithPlatformMatcher(ctx, listImagesOpt...) </mark>
+	if err != nil {
+		return fmt.Errorf("failed to list images: %w", err)
+	}
+	c.loadImages(ctx, cImages)
+...
+...
+}
+
+
+<mark>
+// New functions added in client.go:
+</mark>
+
+func checkIfValidPlatforMatcher(ctx context.Context, provider content.Provider, image ocispec.Descriptor, matcher platforms.MatchComparer) bool {
+	_, err := images.Manifest(ctx, provider, image, matcher)
+	if err != nil {
+		return false
+	}
+	return true
+}
+
+type ImagesWrapper struct {
+	Images Image
+	RuntimeHandler string
+}
+
+func (c *Client) ListImagesWithPlatformMatcher(ctx context.Context, opts ...ListImageOpt) ([]ImagesWrapper, error) {
+	var listImgOpts ListImageOptions
+	for _, o := range opts {
+		if err := o(ctx, &listImgOpts); err != nil {
+			return nil, err
+		}
+	}
+
+	imgs, err := c.ImageService().List(ctx, listImgOpts.Filters...)
+	if err != nil {
+		return nil, err
+	}
+
+	imagesW := make([]ImagesWrapper, len(imgs))
+	if len(opts) == 0 { 
+		for i, img := range imgs {
+			imagesW[i].Images = NewImage(c, img)
+			imagesW[i].RuntimeHandler = c.runtime
+		}
+	} else {
+		for i, img := range imgs {
+			for runtimeHandler, matcher := range listImgOpts.PlatformMatcherMap {
+				tmp := NewImage(c, img)
+				if checkIfValidPlatforMatcher(ctx, tmp.ContentStore(), tmp.Target(), matcher) {
+					imagesW[i].Images = NewImageWithPlatform(c, img, matcher)
+					imagesW[i].RuntimeHandler = runtimeHandler
+				}
+			}
+			if imagesW[i].Images == nil {
+				return nil, fmt.Errorf("valid platform matcher for image %v not found", img)
+			}
+		}
+	}
+	return imagesW, nil
+}
+
+</pre>
+
+
+### Windows platform matcher optional changes
```

Comment:
What is `Windows platform matcher`?

Reply:
Each platform has its own platform matchers defined in https://github.com/containerd/containerd/tree/main/platforms/default*.go files. Windows platform matcher is defined in https://github.com/containerd/containerd/tree/main/platforms/default_windows.go . I will add link to this in KEP for better clarity too.

Reply:
Added more details for this to the section